### PR TITLE
Exercise 2

### PR DESCRIPTION
**Pull request by brickers**
I have deviated from the requirements slightly: rather than hard code a button to move a card to 'Done', the app generates buttons to move the card to other lists. I thought this met the brief, was more useful as you can move to any list you like, and is works with however many lists the user has and whatever they are named. It also satisfied the second part of the brief, letting users move cards back the other way. In a prod environment I would probably create a dropdown menu to tidy up the UI, but haven't done that here.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -49,4 +51,17 @@ You should see output similar to the following:
...
 ```
+
 Now visit [`http://localhost:5000/`](http://localhost:5000/) in your web browser to view the app.
+
+## Trello setup
+
+- Go to Trello.com and sign up for an account, or log in if you already have an account
+- Once signed in, go to https://trello.com/app-key/ and copy the key shown in the gray box at the top. This is your API_KEY.
+- Navigate to https://trello.com/1/authorize?expiration=1day&name=ProjectExercise2&scope=read,write&response_type=token&key={API_KEY}. You will be taken to a page where you are asked to authorise this app to read and write information in your Trello account. Once you click authorise you will be taken to a page with a series of characters shown in red; this is your TOKEN.
+- In the .env file, add two lines under where it says `# Change the following values for local development.`, ensuring the correct copied values are included:
```
Comment (JackMead) [inline README.md:62]: Nice clear instructions!
Minor - but relevant because of just how often people ignore instructions - it would be nice to also add these two keys to the `.env.template` file, in the hope that might catch some people who otherwise miss this part of the instructions

File: todo_app/app.py [hunk 2]
```diff
@@ -1,24 +1,155 @@
...
+import requests
+import os
 
 from todo_app.flask_config import Config
 
 app = Flask(__name__)
 app.config.from_object(Config)
 
 
+if __name__ == '__main__':
```
Comment (JackMead) [inline todo_app/app.py:12]: If we're only ever running the app through `poetry run flask run`, we can actually get rid of this part now

File: todo_app/app.py [hunk 3]
```diff
@@ -1,24 +1,155 @@
...
+            cards = getListCards(list["id"])
+            list["cards"] = cards
+
+        board["lists"] = lists
+        return render_template("board.html", board=board)
+    except:
+        return render_template("error.html")
+
+
+# i would prefer this to be a PATCH route as we are only changing one part of
```
Comment (JackMead) [inline todo_app/app.py:65]: Good point & sensible conclusion, more than happy for this to be POST for exactly this reason

File: todo_app/app.py [hunk 4]
```diff
@@ -1,24 +1,155 @@
...
+# the card object, but HTML forms can only submit GET and POST requests and I
+# didn't want to start writing javascript
+@app.route("/card/<cardId>/list/<listId>", methods=['POST'])
+def moveCardToList(cardId, listId):
+    try:
+        card = getCard(cardId)
+        card.moveToList(listId)
+        if card.lastResponseOK:
```
Comment (JackMead) [inline todo_app/app.py:73]: This is slightly oddly structured - first you use the card id to get the card from trello, then you use the same card id to move it to a list; the first call to Trello only helps us by establishing the board id (but that's not particularly obvious)
Separately, separating the Card class (the info relating to a card) from the logic to move a card about would probably make sense here, e.g. we might make `moveToList` another function similar to `getListCards` etc, that makes the request, rather than asking the card class to manage that itself.

File: todo_app/app.py [hunk 5]
```diff
@@ -1,24 +1,155 @@
...
+            return redirect(f"/board/{card.getBoardID()}")
```
Comment (JackMead) [inline todo_app/app.py:74]: Worth noting that you have an if clause here but no alternative - this view will just error due to no return if the `lastResponseOk` is false

File: todo_app/app.py [hunk 6]
```diff
@@ -1,24 +1,155 @@
...
+
+def getTrelloItem(url):
+    response = requests.get(url, headers=auth_header)
+    if response.ok:
+        return response.json()
+
+
+class Card:
+    def __init__(self, *args):
+        if len(args) > 1 and isinstance(args[0], str) and isinstance(args[1], str):
```
Comment (JackMead) [inline todo_app/app.py:121]: This is reasonably hard to follow, an alternative pattern would be to use a static factory method to split that out, e.g.
```

class Card:
    def __init__(self, name, idList, idCard = None, idBoard = None, ):
        self.name = name
        self.idList = idList
        if idCard == None or idBoard == None:
            url = BASE_URL + f"/1/cards"
            params = {
                "name": self.name,
                "idList": self.idList
            }
            response = requests.post(url, params=params, headers=auth_header)
            json =  response.json()
            self.idBoard = json['idBoard']
            self.id = json['id']
            self.lastResponseOK = response.ok
        else:
            self.id = idCard
            self.idBoard = idBoard

    @classmethod
    def from_trello_card(cls, trello_response):
        json = trello_response
        name = json["name"]
        id = json["id"]
        idList = json["idList"]
        idBoard = json["idBoard"]
        cls(name, id, idList, idBoard)
```

Saying that, we could certainly improve this further if we separated the Card class from the logic inside it that creates a card - we'd then only need a single constructor for now (since addCardToList wouldn't need to actually create a card)

File: todo_app/app.py [hunk 7]
```diff
@@ -1,24 +1,155 @@
...
+    try:
+        board = getBoard(id)
+        lists = getBoardLists(id)
+
+        for list in lists:
+            cards = getListCards(list["id"])
+            list["cards"] = cards
+
+        board["lists"] = lists
+        return render_template("board.html", board=board)
```
Comment (JackMead) [inline todo_app/app.py:60]: Here we're passing through essentially the json directly from Trello  (albeit modified to combine some useful fields), rather than creating our own class to use in the templates

**Review by JackMead — CHANGES_REQUESTED** [review]
This looks good, more than happy with your decision re the buttons to change lists, and I like the feature to search for different boards to start with!
I've left a few thoughts below, but the main thing we should adjust is that the class added for Part 5 (`Card` in your case) is a little confused, the idea here is to use the class when passed through to the templates, and it should know how to digest a trello response (as you have it) but it shouldn't need to make calls to trello within the class itself - I wouldn't expect to see that class required when creating a new card or moving a card between lists

Feel free to shout if that's not clear, (or indeed even if the motivation for it isn't!)

File: todo_app/app.py (hunk 5 above)
Comment (brickers) [reply todo_app/app.py:74]: Hi Jack. I'm not totally sure I considered this, and if I did then perhaps I assumed that because it's in a `try` block it would then be managed by the exception, but I see now that no error occurs so that probably wouldn't happen. I will update with an `else` leading to the error page.

File: todo_app/app.py (hunk 7 above)
Comment (brickers) [reply todo_app/app.py:55]: this has been changed now

File: todo_app/app.py (hunk 6 above)
Comment (brickers) [reply todo_app/app.py]: thanks for this. i think i was trying to do something I've done in other languages and really it wasn't clear. certainly i found it hard to read coming back to it! also good to see how the confusion over the purpose of Class caused this to be more complex than necessary

File: todo_app/app.py (hunk 4 above)
Comment (brickers) [reply todo_app/app.py]: sorted
